### PR TITLE
Preventing sending whole list of application again and again.

### PR DESCRIPTION
**Pull request by shagihan**
Preventing sending the whole list of the application again and again.

### CODE REVIEW

File: client/client/src/main/java/org/wso2/emm/agent/services/operation/OperationManager.java [hunk 1]
```diff
@@ -268,6 +270,34 @@ public void getApplicationList(org.wso2.emm.agent.beans.Operation operation) thr
         JSONArray result = new JSONArray();
         RuntimeInfo runtimeInfo = new RuntimeInfo(context);
         Map<String, Application> applications = runtimeInfo.getAppMemory();
+        Gson gson = new Gson();
+        JsonElement element = null;
+        if (Preference.getString(context, Constants.LAST_APP_LIST_SHARED_PREF) == null) {
+            element = gson.toJsonTree(apps, new TypeToken<ArrayList<DeviceAppInfo>>() {
+            }.getType());
+            if (!element.isJsonArray()) {
+                throw new AndroidAgentException("Invalid JSON format. Not JSON array.");
```
Comment (charithag) [inline client/client/src/main/java/org/wso2/emm/agent/services/operation/OperationManager.java]: Also include json string which leads to this exception

File: client/client/src/main/java/org/wso2/emm/agent/services/operation/OperationManager.java [hunk 2]
```diff
@@ -268,6 +270,34 @@ public void getApplicationList(org.wso2.emm.agent.beans.Operation operation) thr
...
+            }
+            String applicationList = element.getAsJsonArray().toString().replace(" ", "");
+            Preference.putString(context, Constants.LAST_APP_LIST_SHARED_PREF, applicationList);
+        } else {
+            String previousAppList = Preference.getString(context, Constants.LAST_APP_LIST_SHARED_PREF);
+            element = gson.toJsonTree(apps, new TypeToken<ArrayList<DeviceAppInfo>>() {
+            }.getType());
+            if (!element.isJsonArray()) {
+                throw new AndroidAgentException("Invalid JSON format. Not JSON array.");
```
Comment (charithag) [inline client/client/src/main/java/org/wso2/emm/agent/services/operation/OperationManager.java]: Also include json string which leads to this exception

**Review by charithag — CHANGES_REQUESTED** [review]

File: client/client/src/main/java/org/wso2/emm/agent/services/operation/OperationManager.java (hunk 1 above)
Comment (shagihan) [reply client/client/src/main/java/org/wso2/emm/agent/services/operation/OperationManager.java]: Invalid JSON appended to the exception.

File: client/client/src/main/java/org/wso2/emm/agent/services/operation/OperationManager.java (hunk 2 above)
Comment (shagihan) [reply client/client/src/main/java/org/wso2/emm/agent/services/operation/OperationManager.java]: Invalid JSON appended to the exception.

**Review by charithag — APPROVED** [review]

File: client/client/src/main/java/org/wso2/emm/agent/services/DeviceStartupIntentReceiver.java [hunk 3]
```diff
@@ -111,6 +111,9 @@ private void setRecurringAlarm(Context context) {
         if (Preference.getString(context, Constants.LAST_WIFI_SCAN_RESULT_SHARED_PREF) != null) {
             Preference.putString(context, Constants.LAST_WIFI_SCAN_RESULT_SHARED_PREF, null);
         }
+        if(Preference.getString(context, Constants.LAST_APP_LIST_SHARED_PREF) != null) {
```
Comment (hasuniea) [inline client/client/src/main/java/org/wso2/emm/agent/services/DeviceStartupIntentReceiver.java]: Could you please format the code

Comment (shagihan) [reply client/client/src/main/java/org/wso2/emm/agent/services/DeviceStartupIntentReceiver.java]: Resolved.

**Review by hasuniea — APPROVED** [review]

**Pull request merged**